### PR TITLE
Makeflow Mesos Support

### PR DESCRIPTION
**Pull request by charleszheng44**
This PR is for mesos mode, which rebased the latest master branch.

### CODE REVIEW

**Comment by dthain** [comment]
I see that you added a few lines to `makeflow.html` indicating that there exists a mesos mode, but I think this is going to be lost in the shuffle.  Please add a new subsection to the docs indicating the Mesos mode, how to set it up, any special considerations, etc...  That will make it more likely for people to find it and use it.

**Comment by dthain** [comment]
@btovar @nhazekam Do you see any problems here wrt the core design of Makeflow?  (It's ok if the details of Mesos are in flux, as long as core Makeflow is not broken.)

File: makeflow/src/makeflow.c [hunk 1]
```diff
@@ -2046,14 +2078,13 @@ if (enforcer && wrapper_umbrella) {
...
-
 	runtime = timestamp_get();
 
 	if (container_mode == CONTAINER_MODE_DOCKER) {
-            makeflow_wrapper_docker_init(wrapper, container_image, image_tar);
+    	makeflow_wrapper_docker_init(wrapper, container_image, image_tar);
 	}else if(container_mode == CONTAINER_MODE_SINGULARITY){
-            makeflow_wrapper_singularity_init(wrapper, container_image);
-        }
+    	makeflow_wrapper_singularity_init(wrapper, container_image);
```
Comment (nhazekam) [inline makeflow/src/makeflow.c]: This does not affect the software working, but the spacing here is messed up.

File: makeflow/src/makeflow.c [hunk 2]
```diff
@@ -2065,6 +2096,21 @@ if (enforcer && wrapper_umbrella) {
...
+	/*
+	 * Set the abort and failed flag for batch_job_mesos mode.
+	 * Since batch_queue_delete(struct batch_queue *q) will call
+	 * batch_queue_mesos_free(struct batch_queue *q), which is defined 
+	 * in batch_job/src/batch_job_mesos.c. Then this function will check 
+	 * the abort and failed status of the batch_queue and inform 
+	 * the makeflow mesos scheduler. 
+	 */
+
+	if (batch_queue_type == BATCH_QUEUE_TYPE_MESOS) {
```
Comment (nhazekam) [inline makeflow/src/makeflow.c]: Good point on letting Mesos know about failure, but is there any reason this needs to be Mesos specific? Does not need to be changed, but curious is this is useful information for other batch-queues.

File: makeflow/src/mf_mesos_executor [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    def __init__(self, cmd, executor_id, framework_id, hostname):
+        self.hostname = hostname
+        self.port = DEFAULT_PORT
+        self.cmd = cmd
+        self.executor_id = executor_id
+        self.framework_id = framework_id
+
+    def disconnected(self, driver):
+        driver.sendFrameworkMessage("[EXUT_STATE] {0} {1} disconnected".format(self.executor_id, self.task_id))
```
Comment (nhazekam) [inline makeflow/src/mf_mesos_executor]: Just clarifying this shouldn't be EXIT_STATE.

File: resource_monitor/src/resource_monitor_histograms.c [hunk 4]
```diff
@@ -362,7 +362,7 @@ void write_variables_gnuplot(struct field_stats *h, struct field_stats *all)
 	fprintf(f, "%s = %lf\n", "current_first_allocation_min_waste", rmsummary_to_external_unit(h->field, h->fa_min_waste_time_dependence.first));
 
 	fprintf(f, "%s = %lf\n", "current_bin_size",   rmsummary_to_external_unit(h->field, histogram_bucket_size(h->histogram)));
-
+
```
Comment (nhazekam) [inline resource_monitor/src/resource_monitor_histograms.c]: Unnecessary Change.

File: batch_job/src/batch_job.h [hunk 5]
```diff
@@ -44,6 +44,7 @@ typedef enum {
 	BATCH_QUEUE_TYPE_CLUSTER,             /**< Batch jobs will be sent to a user-defined cluster manager. */
 	BATCH_QUEUE_TYPE_WORK_QUEUE,          /**< Batch jobs will be sent to the Work Queue. */
 	BATCH_QUEUE_TYPE_CHIRP,               /**< Batch jobs will be sent to Chirp. */
+    BATCH_QUEUE_TYPE_MESOS,               /**< Batch jobs will be sent to Mesos. */
```
Comment (nhazekam) [inline batch_job/src/batch_job.h]: Spacing looks off.

File: makeflow/src/makeflow.c [hunk 6]
```diff
@@ -1851,6 +1876,13 @@ if (enforcer && wrapper_umbrella) {
 		}
 	}
 
+	if(batch_queue_type == BATCH_QUEUE_TYPE_MESOS) {
+		batch_queue_set_option(remote_queue, "mesos-path", mesos_path);
+		batch_queue_set_option(remote_queue, "mesos-master", mesos_master);
+		batch_queue_set_option(remote_queue, "mesos-preload", mesos_preload);
+		batch_queue_set_feature(remote_queue, "batch_log_name", batchlogfilename);
```
Comment (nhazekam) [inline makeflow/src/makeflow.c]: Is setting the batch log name specific to Mesos? This should be covered by the if statement just previous using the feature you set in batch_job_mesos.c

File: dttools/src/text_list.c [hunk 7]
```diff
@@ -40,6 +40,24 @@ struct text_list *text_list_load(const char *path)
...
+struct text_list *text_list_load_str(const char *inp_str)
+{
+	if (!inp_str) {
+		return NULL;
+	}
+
+	struct text_list *t = text_list_create();
+	char *pch = NULL;
+	char * tmp_str = strdup(inp_str);
+    pch = strtok(tmp_str, ",");
```
Comment (nhazekam) [inline dttools/src/text_list.c]: Nit-pick spacing issue.

File: batch_job/src/work_queue_factory.c [hunk 8]
```diff
@@ -108,6 +107,10 @@ static int64_t factory_timeout = 0;
 
 struct batch_queue *queue = 0;
 
+static const char *mesos_master = NULL;
```
Comment (nhazekam) [inline batch_job/src/work_queue_factory.c]: I believe these three variable are only used in main. Do they need to be static variables?

File: doc/makeflow.html [hunk 9]
```diff
@@ -329,6 +330,25 @@ <h2 id="umbrella_in_makefile">Specify Umbrella in Makefile<a class="sectionlink"
...
 wrapped by Umbrella. </p>
 
+<h2 id="mesos">Running Makeflow with Mesos<a class="sectionlink" href="#mesos" title="Link to this section.">&#x21d7;</a></h2>
+
+<p>Makeflow can be used with Apache Mesos. To run Makeflow with Mesos, user need to 
+specify the batch mode via <tt> -T mesos </tt> option and pass the hostname and 
+port number of Mesos master to Makeflow through <tt> --mesos-master </tt> option.
+Since the Makeflow-Mesos Scheduler is based on Mesos Python2 API, the path to Mesos 
+Python2 library should be included in the <tt>$PATH</tt>, or one can specify a 
+preferrd Mesos Python2 API via <tt> --mesos-path </tt> option. To successfully
```
Comment (nhazekam) [inline doc/makeflow.html]: preferred -  spelling

File: doc/makeflow.html [hunk 10]
```diff
@@ -329,6 +330,25 @@ <h2 id="umbrella_in_makefile">Specify Umbrella in Makefile<a class="sectionlink"
...
+specify the batch mode via <tt> -T mesos </tt> option and pass the hostname and 
+port number of Mesos master to Makeflow through <tt> --mesos-master </tt> option.
+Since the Makeflow-Mesos Scheduler is based on Mesos Python2 API, the path to Mesos 
+Python2 library should be included in the <tt>$PATH</tt>, or one can specify a 
+preferrd Mesos Python2 API via <tt> --mesos-path </tt> option. To successfully 
+import the Python library, user may need to indicate dependencies through 
+<tt> --mesos-preload </tt> option. </p>
+
+<p>Following is the command to run Makeflow on a Mesos cluster that has the master listen 
+on port 5050 of localhost, with a user specified python library that dependents on an
```
Comment (nhazekam) [inline doc/makeflow.html]: is dependent on  --or-- depends on 
Odd wording.

File: makeflow/src/makeflow.c (hunk 1 above)
Comment (charleszheng44) [inline makeflow/src/makeflow.c]: @nhazekam What tool are we using for dealing with tab and whitespace? I have been bothered by this problem by a long time.

File: batch_job/src/batch_job.c [hunk 11]
```diff
@@ -42,7 +43,7 @@ static struct batch_queue_module batch_queue_unknown = {
 	{NULL, NULL, NULL, NULL, NULL, NULL},
 };
 
-#define BATCH_JOB_SYSTEMS  "local, wq, condor, sge, torque, moab, slurm, chirp, amazon, dryrun"
+#define BATCH_JOB_SYSTEMS  "local, wq, mesos, condor, sge, torque, moab, slurm, chirp, amazon, dryrun"
```
Comment (btovar) [inline batch_job/src/batch_job.c]: Let's have these ones in alphabetical order.

File: batch_job/src/batch_job_mesos.c [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+		if (dup2(mesos_fd, 2) == -1) {
+	    	debug(D_ERROR, "%s\n", strerror(errno));
+	    	exit(EXIT_FAILURE);
+	   	}
+
+	    close(mesos_fd);
+
+		execle("/usr/bin/python", "python", exe_py_path, mesos_cwd, mesos_master, mesos_python_path, (char *) 0, envs);
+
+		_exit(127);
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: No need for _. Why 127? Better:  exit(errno);

File: batch_job/src/batch_job_mesos.c [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+static const char *mesos_master = NULL;
+static const char *mesos_preload = NULL;
+static pid_t mesos_scheduler_pid = 0;
+
+static void start_mesos_scheduler(struct batch_queue *q)
+{
+
+	pid_t mesos_pid;
+	mesos_pid = fork();			
+	mesos_scheduler_pid = mesos_pid;
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: Only relevant to fork parent.

File: batch_job/src/batch_job_mesos.c [hunk 14]
```diff
@@ -0,0 +1,437 @@
...
+	char *mesos_cwd;
+	mesos_cwd = path_getcwd();
+
+	char exe_path[MAX_BUF_SIZE];
+
+	if(readlink("/proc/self/exe", exe_path, MAX_BUF_SIZE) == -1) {
+		debug(D_ERROR, "read \"proc/self/exe\" fail\n");
+		exit(EXIT_FAILURE);
+	}
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: Only relevant to fork child. Move to corresponding sections of if else below, to make the code clear.

File: batch_job/src/batch_job_mesos.c [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+	    close(mesos_fd);
+
+		execle("/usr/bin/python", "python", exe_py_path, mesos_cwd, mesos_master, mesos_python_path, (char *) 0, envs);
+
+		_exit(127);
+
+	} else {
+
+		debug(D_ERROR, "mesos batch system couldn't create new process: %s\n", strerror(errno));
+		exit(EXIT_FAILURE);
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: exit(errno)

File: batch_job/src/batch_job_mesos.c [hunk 16]
```diff
@@ -0,0 +1,437 @@
...
+
+	}
+
+}
+
+
+static batch_job_id_t batch_job_mesos_submit (struct batch_queue *q, const char *cmd, \
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: Do not break lines with \\. It is not python...

File: batch_job/src/batch_job_mesos.c [hunk 17]
```diff
@@ -0,0 +1,437 @@
...
+		const char *extra_input_files, const char *extra_output_files, \
+		struct jx *envlist, const struct rmsummary *resources )
+{
+
+	// Get the path to mesos python site-packages
+	if (is_mesos_py_path_known != 1) {
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: Usually we say:   if(!is_mesos...),  given that is_mesos... is zero or anything else.

File: batch_job/src/batch_job_mesos.c [hunk 18]
```diff
@@ -0,0 +1,437 @@
...
+		mesos_py_path = batch_queue_get_option(q, "mesos-path");
+		if (mesos_py_path == NULL) {
+			debug(D_ERROR, "Please set the path to mesos python library\n");
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: How?

File: batch_job/src/batch_job_mesos.c [hunk 19]
```diff
@@ -0,0 +1,437 @@
...
+			exit(EXIT_FAILURE);
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: For such fatal errors, we combine the debug and exit in a single line:

fatal('Please set the path ....)

File: batch_job/src/batch_job_mesos.c [hunk 20]
```diff
@@ -0,0 +1,437 @@
...
+		} else {
+			debug(D_INFO, "Get mesos_path %s from command line\n", mesos_py_path);
+			is_mesos_py_path_known = 1;
+		}
+	}
+
+	// Get the mesos master address
+	if (is_mesos_master_known != 1) {
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: As above.

File: batch_job/src/batch_job_mesos.c [hunk 21]
```diff
@@ -0,0 +1,437 @@
...
+			exit(EXIT_FAILURE);
+		} else {
+			debug(D_INFO, "Get mesos_path %s from command line\n", mesos_py_path);
+			is_mesos_master_known = 1;
+		}
+	}
+
+	mesos_preload = batch_queue_get_option(q, "mesos-preload");
+
+	if (is_mesos_py_path_known == 1 && \
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: Do not use \.

Better:  if(is_mesos_py...   && is_mesos_master ...  && !is_scheduler_running) {

File: makeflow/src/makeflow.c (hunk 1 above)
Comment (nhazekam) [inline makeflow/src/makeflow.c]: Unfortunately for spacing like this I have tabs equal to 4 spaces in vim and just match the current file, so particular tool. For whitespace, like tabs, I usually just remove the changes, though @batrick explained a git method for removing extraneous whitespace. I will try to find that conversations.

File: batch_job/src/batch_job_mesos.c [hunk 22]
```diff
@@ -0,0 +1,437 @@
...
+		// start mesos scheduler if it is not running
+		start_mesos_scheduler(q);
+		is_scheduler_running = 1;
+	}
+
+	int task_id = ++counter;
+
+	debug(D_BATCH, "task %d is ready", task_id);
+	struct batch_job_info *info = malloc(sizeof(*info));
+	memset(info, 0, sizeof(*info));
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: Combine malloc and memset 0 with:

... *info = calloc(1, sizeof(*info));

File: batch_job/src/batch_job_mesos.c [hunk 23]
```diff
@@ -0,0 +1,437 @@
...
+	info->submitted = time(0);
+	info->started = time(0);
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: info->started = time(0);
info->submitted = info->started;

File: batch_job/src/batch_job_mesos.c [hunk 24]
```diff
@@ -0,0 +1,437 @@
...
+	itable_insert(q->job_table, task_id, info);
+
+	FILE *task_info_fp;
+
+	if(access(FILE_TASK_INFO, F_OK) != -1) {
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: Add a comment of why this works.

File: batch_job/src/batch_job_mesos.c [hunk 25]
```diff
@@ -0,0 +1,437 @@
...
+
+		int j = 0;
+		int num_input_files = text_list_size(mt->task_input_files);
+		for (j = 0; j < (num_input_files-1); j++) {
+			fprintf(task_info_fp, "%s ", text_list_get(mt->task_input_files, j));
+		}
+		fprintf(task_info_fp, "%s,", text_list_get(mt->task_input_files, num_input_files-1));
+
+	} else {
+		fputs(",", task_info_fp);
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: fprintf(task_info_fp, ",");

File: batch_job/src/batch_job_mesos.c [hunk 26]
```diff
@@ -0,0 +1,437 @@
...
+    
+	if (extra_output_files != NULL && strlen(extra_output_files) != 0) {
+		int j = 0;
+		int num_output_files = text_list_size(mt->task_output_files);
+		for (j = 0; j < (num_output_files-1); j++) {
+			fprintf(task_info_fp, "%s ", text_list_get(mt->task_output_files, j));
+		}
+		fprintf(task_info_fp, "%s,",text_list_get(mt->task_output_files, num_output_files-1));
+	} else {
+		fputs(",", task_info_fp);
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: As above.

File: batch_job/src/batch_job_mesos.c [hunk 27]
```diff
@@ -0,0 +1,437 @@
...
+
+static batch_job_id_t batch_job_mesos_wait (struct batch_queue * q, struct batch_job_info * info_out, time_t stoptime)
+{
+		
+	char *line = NULL;
+	size_t len = 0;
+	ssize_t read_len;
+	FILE *task_state_fp;
+
+	if(finished_tasks == NULL) {
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: if(!finished_tasks) {

File: batch_job/src/batch_job_mesos.c [hunk 28]
```diff
@@ -0,0 +1,437 @@
...
+			    	
+				info->finished = time(0);
+				task_stat_str = strtok(NULL, ",");
+
+				if (strcmp(task_stat_str, "finished") == 0) {
+					info->exited_normally = 1;
+				} else if (strcmp(task_stat_str, "failed") == 0) {
+					info->exited_normally = 0;
+					task_exit_code = strtok(NULL, ",");
+					if(atoi(task_exit_code) == 444) {
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: Comment... .where did the 444 come from?

File: batch_job/src/batch_job_mesos.c [hunk 29]
```diff
@@ -0,0 +1,437 @@
...
+
+		if(stoptime != 0 && time(0) >= stoptime) {
+			fclose(task_state_fp);
+			return -1;
+		}
+	}
+
+}
+
+static int batch_job_mesos_remove (struct batch_queue *q, batch_job_id_t jobid)
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: This function makes me queasy. The timeout = 40, and the sleep for one second  seem magical. 
Why is a timeout needed at all?
What happens to the lines not processed when the timeout is hit?

File: batch_job/src/batch_job_mesos.c [hunk 30]
```diff
@@ -0,0 +1,437 @@
...
+	if(fp == NULL) {
+		debug(D_ERROR, "Fail to clean up batch queue. %s\n", strerror(errno)); 
+		return 1;
+	}
+
+	int batch_queue_abort_flag = atoi(batch_queue_get_option(q, "batch-queue-abort-flag"));
+	int batch_queue_failed_flag = atoi(batch_queue_get_option(q, "batch-queue-failed-flag"));
+
+	if(batch_queue_abort_flag) {
+		fputs("aborted", fp);
```
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: Don't use fputs. Transformed to fprintf(fp, "...")

File: batch_job/src/mesos_task.c [hunk 31]
```diff
@@ -0,0 +1,43 @@
+#include "mesos_task.h"
+
+struct mesos_task *mesos_task_create(int task_id, const char *cmd, \
```
Comment (btovar) [inline batch_job/src/mesos_task.c]: No backslashes.

File: batch_job/src/work_queue_factory.c (hunk 8 above)
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: I believe nick means if they need to be static global, rather than regular local to main.

File: batch_job/src/batch_job_mesos.c (hunk 28 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: I set 444 on mesos side. in `makeflow/src/mf_mesos_scheduler`

File: doc/makeflow.html (hunk 9 above)
Comment (charleszheng44) [inline doc/makeflow.html]: change preferrd to preferred

File: doc/makeflow.html (hunk 10 above)
Comment (charleszheng44) [inline doc/makeflow.html]: change to depends

File: batch_job/src/batch_job.c (hunk 11 above)
Comment (charleszheng44) [inline batch_job/src/batch_job.c]: change to "... mesos, moab..."

File: batch_job/src/batch_job_mesos.c (hunk 12 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: change to exit(errno)

File: batch_job/src/batch_job_mesos.c (hunk 15 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: change to exit(errno)

File: batch_job/src/batch_job_mesos.c (hunk 16 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: good to know! I do not like '\' ...

File: batch_job/src/batch_job_mesos.c (hunk 17 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: change to `if(!is_mesos...)`

File: batch_job/src/batch_job_mesos.c (hunk 20 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: change to `if(!is_mesos_master_known)`

File: batch_job/src/batch_job_mesos.c (hunk 21 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: Change to `if(is_mesos_py... && is_mesos_master ... && !is_scheduler_running)`

File: batch_job/src/batch_job_mesos.c (hunk 22 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: Use `calloc` instead of `malloc + memset`

File: batch_job/src/batch_job_mesos.c (hunk 23 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: changed to 
`
info->started = time(0);
info->submitted = info->started;
`

File: batch_job/src/batch_job_mesos.c (hunk 25 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: change to `fprintf`

File: batch_job/src/batch_job_mesos.c (hunk 26 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: change to `fprintf`

File: batch_job/src/batch_job_mesos.c (hunk 27 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: change to `if(!finished_tasks)`

File: batch_job/src/batch_job_mesos.c (hunk 30 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: change to `fprintf`

File: batch_job/src/mesos_task.c (hunk 31 above)
Comment (charleszheng44) [inline batch_job/src/mesos_task.c]: delete backslashes!

File: batch_job/src/batch_job_mesos.c (hunk 19 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: change to `fatal`

File: batch_job/src/batch_job_mesos.c (hunk 18 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: add more detailed error message

File: batch_job/src/batch_job_mesos.c (hunk 13 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: Remove useless variable

File: batch_job/src/batch_job_mesos.c (hunk 14 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: Move variables that only relevant to child process to corresponding code section

File: batch_job/src/work_queue_factory.c (hunk 8 above)
Comment (charleszheng44) [inline batch_job/src/work_queue_factory.c]: I see. I noticed that there are many variables are defined as `static char *` and only used once in `main`, like (`amazon_credentials`, `amazon_ami`, `wrapper_command`, etc), is this a proper way of define variables?

File: batch_job/src/batch_job_mesos.c (hunk 29 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: The timeout is for checking if mesos successfully stop the container in the cluster in 40 seconds. If the timeout is hit, then we will say the `batch_job_mesos_remove` failed. Since it is hard to say if the mesos can stop the container in 40 seconds, I just disabled the timeout.

File: batch_job/src/work_queue_factory.c (hunk 8 above)
Comment (nhazekam) [inline batch_job/src/work_queue_factory.c]: I agree. For clarity, you are creating this variable and passing it to the batch_job system. The variable from makeflow's perspective is not important, it only means something in batch_job. The others that you mentioned should also be cleaned up, though I am not relying on you for that.

File: makeflow/src/mf_mesos_executor (hunk 3 above)
Comment (charleszheng44) [inline makeflow/src/mf_mesos_executor]: EXUT_STATE means executor state

File: resource_monitor/src/resource_monitor_histograms.c (hunk 4 above)
Comment (charleszheng44) [inline resource_monitor/src/resource_monitor_histograms.c]: revert this file to previous version

File: makeflow/src/mf_mesos_executor (hunk 3 above)
Comment (btovar) [inline makeflow/src/mf_mesos_executor]: Is that a mesos thing? If not, that will look like a typo to people in the future.  EXECUTOR_STATE is better.

Comment (charleszheng44) [inline makeflow/src/mf_mesos_executor]: change EXUT_STATE to EXECUTOR_STATE

File: makeflow/src/makeflow.c (hunk 2 above)
Comment (charleszheng44) [inline makeflow/src/makeflow.c]: Since Mesos run each task with an independent executor, if user abort a workflow that contains many long running tasks, it'd be better to notice Mesos to stop executors work for this workflow.  I am not sure how do other batch job systems abort running tasks, but maybe we should notice other batch job systems about failure as well.

File: batch_job/src/work_queue_factory.c (hunk 8 above)
Comment (charleszheng44) [inline batch_job/src/work_queue_factory.c]: Change static global variables to local main variables

File: batch_job/src/batch_job_mesos.c (hunk 24 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: add explanation.

**Comment by charleszheng44** [comment]
@btovar @dthain @nhazekam  I think this PR is ready to be merged.

File: batch_job/src/batch_job_mesos.c (hunk 28 above)
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: Please add a comment that the 444 is arbitrary, set on makeflow/src/mf_mesos_scheduler, and under which condition we expect it (e.g., missing output?)

File: batch_job/src/batch_job_mesos.c (hunk 29 above)
Comment (btovar) [inline batch_job/src/batch_job_mesos.c]: So what happens now? we wait forever for mesos? (I'm not judging if it is right or wrong now, I just want you to include comments on what the code is doing.)

Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: To remove a batch job, we mark the task state as "aborting", and return. Then 
the mesos scheduler will try to terminate the corresponding executors. This 
method does not guarantee the termination of executors, but all executors would be
terminated before the mesos scheduler stop.

Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: I have added the comment.

File: batch_job/src/batch_job_mesos.c (hunk 28 above)
Comment (charleszheng44) [inline batch_job/src/batch_job_mesos.c]: A comment is added to explain the exit code.

**Comment by dthain** [comment]
@btovar can you check please?
